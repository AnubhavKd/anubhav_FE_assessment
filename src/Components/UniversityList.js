import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import './styles/UniversityList.css'; 

const UniversityList = ({ universities, onDelete }) => {
  const initialSearchTerm = localStorage.getItem('searchTerm') || '';
  const initialSortBy = localStorage.getItem('sortBy') || '';

  const [searchTerm, setSearchTerm] = useState(initialSearchTerm);
  const [sortBy, setSortBy] = useState(initialSortBy);


  console.log("university list", universities)

  useEffect(() => {
    localStorage.setItem('searchTerm', searchTerm);
  }, [searchTerm]);

  useEffect(() => {
    localStorage.setItem('sortBy', sortBy);
  }, [sortBy]);

  const handleSearchChange = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleSortByChange = (event) => {
    setSortBy(event.target.value);
  };

  const handleDelete = (name) => {
    const confirmDelete = window.confirm(`Are you sure you want to delete ${name}?`);

    if (confirmDelete) {
      onDelete(name);
    }
  };

  const filteredUniversities = universities.filter((university) =>
    university.name.toLowerCase().includes(searchTerm.toLowerCase())
  );

  let sortedUniversities = [...filteredUniversities];

  if (sortBy === 'asc') {
    sortedUniversities.sort((a, b) => a.name.localeCompare(b.name));
  } else if (sortBy === 'desc') {
    sortedUniversities.sort((a, b) => b.name.localeCompare(a.name));
  }

  return (
    <div className="container">
      <h2 data-testid="university-list-title">University List</h2>
      <div className="filter-sort">
        <input
          type="text"
          placeholder="Search by university name..."
          value={searchTerm}
          onChange={handleSearchChange}
          className="search-input"
        />
        <select value={sortBy} onChange={handleSortByChange} className="sort-select">
          <option value="">Default</option>
          <option value="asc">A-Z</option>
          <option value="desc">Z-A</option>
        </select>
      </div>
      {sortedUniversities.length === 0 ? (
        <p>No universities found.</p>
      ) : (
        sortedUniversities.map((university, index) => (
          <div key={index} className="university-card">
            <h3>{university.name}</h3>
            <p>Country: {university.country}</p>
            {university.state_province && <p>State/Province: {university.state_province}</p>}
            <p>
              Website: <a href={university.web_pages[0]} target="_blank" rel="noopener noreferrer">{university.web_pages[0]}</a>
            </p>
            <button className="delete_button" onClick={() => handleDelete(university.name)}>Delete</button>

            <button className="link_button">
              <Link to={`/university/${encodeURIComponent(university.name)}`}>View Details</Link>
            </button>
          </div>
        ))
      )}
    </div>
  );
};

export default UniversityList;

