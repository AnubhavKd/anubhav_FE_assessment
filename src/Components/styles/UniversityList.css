
.container {
    max-width: 800px;
    margin: 0 auto;
    padding: 20px;
  }
  
  h2 {
    font-size: 24px;
    margin-bottom: 20px;
  }
  
  .filter-sort {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 20px;
  }
  
  .search-input,
  .sort-select {
    width: 45%;
    padding: 10px;
    font-size: 16px;
    margin-right: 10px;
  }
  
  .university-card {
    border: 1px solid #ddd;
    border-radius: 8px;
    padding: 20px;
    margin-bottom: 20px;
    background-color: #ffffff;
  }
  
  h3 {
    font-size: 20px;
    margin-bottom: 10px;
  }
  
  p {
    font-size: 16px;
    margin-bottom: 8px;
  }
  
  a {
    color: #007bff;
  }
  
  .delete_button {
    padding: 8px 16px;
    margin-top: 10px;
    background-color: #C00001;
    color: #fff;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    margin-right: 10px;
  }

  .link_button {
    padding: 8px 16px;
    font-size: 16px;
    margin-top: 10px;
    color: #0056b3; /* Default text color */
    background-color: #ffffff;
    border: none;
    border-radius: 4px;
    margin-right: 10px;
}



