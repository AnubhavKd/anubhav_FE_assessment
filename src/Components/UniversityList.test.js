import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom'; 
import UniversityList from './UniversityList';

describe('UniversityList Component', () => {
  const universities = [
    {
      name: 'Harvard University',
      country: 'USA',
      state_province: 'Massachusetts',
      web_pages: ['http://www.harvard.edu']
    },
    {
      name: 'Oxford University',
      country: 'UK',
      web_pages: ['http://www.ox.ac.uk']
    }
  ];

  const onDelete = jest.fn(); // Mock delete function

  it('renders university list component with title', () => {
    render(
      <Router>
        <UniversityList universities={universities} onDelete={onDelete} />
      </Router>
    );

    const titleElement = screen.getByTestId('university-list-title');
   

    
    expect(titleElement).toBeInTheDocument();
    expect(titleElement).toHaveTextContent('University List');
  });


  it('renders university list component with list of universities', () => {
    render(
      <Router>
        <UniversityList universities={universities} onDelete={onDelete} />
      </Router>
    );

    const getByRole=screen.getByRole('textbox')
    expect (getByRole).toBeInTheDocument()


  })

  it('renders sorting alphabetical order of list', () => {
    render(
      <Router>
        <UniversityList universities={universities} onDelete={onDelete} />
      </Router>
    );
    const getSortingFilter=screen.getByRole('combobox')
    expect (getSortingFilter).toBeInTheDocument()
  })


  it('testing correct title of university list', () => {
    render(
      <Router>
        <UniversityList universities={universities} onDelete={onDelete} />
      </Router>
    );
    const title=screen.getByRole('heading', {
        name: /university list/i
      })
    expect (title).toBeInTheDocument()
    
  })


  it('renders UniversityList component correctly', () => {
    const { asFragment } = render(
      <Router>
        <UniversityList universities={universities} onDelete={onDelete} />
      </Router>
    );

    // Use asFragment() to retrieve the component's rendered output as a snapshot
    expect(asFragment()).toMatchSnapshot();
  });


 




 

});
