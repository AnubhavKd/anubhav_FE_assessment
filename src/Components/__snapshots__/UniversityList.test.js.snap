// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniversityList Component renders UniversityList component correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h2
      data-testid="university-list-title"
    >
      University List
    </h2>
    <div
      class="filter-sort"
    >
      <input
        class="search-input"
        placeholder="Search by university name..."
        type="text"
        value=""
      />
      <select
        class="sort-select"
      >
        <option
          value=""
        >
          Default
        </option>
        <option
          value="asc"
        >
          A-Z
        </option>
        <option
          value="desc"
        >
          Z-A
        </option>
      </select>
    </div>
    <div
      class="university-card"
    >
      <h3>
        Harvard University
      </h3>
      <p>
        Country: USA
      </p>
      <p>
        State/Province: Massachusetts
      </p>
      <p>
        Website: 
        <a
          href="http://www.harvard.edu"
          rel="noopener noreferrer"
          target="_blank"
        >
          http://www.harvard.edu
        </a>
      </p>
      <button
        class="delete_button"
      >
        Delete
      </button>
      <button
        class="link_button"
      >
        <a
          href="/university/Harvard%20University"
        >
          View Details
        </a>
      </button>
    </div>
    <div
      class="university-card"
    >
      <h3>
        Oxford University
      </h3>
      <p>
        Country: UK
      </p>
      <p>
        Website: 
        <a
          href="http://www.ox.ac.uk"
          rel="noopener noreferrer"
          target="_blank"
        >
          http://www.ox.ac.uk
        </a>
      </p>
      <button
        class="delete_button"
      >
        Delete
      </button>
      <button
        class="link_button"
      >
        <a
          href="/university/Oxford%20University"
        >
          View Details
        </a>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
